File: Java/QuickSort.java
public class QuickSort {
  /**
   * QuickSort v.1
   * it uses naive, partition, swap
   */
  public void naive(int[] fin, int low, int high) {
    if (low >= high)
      return;

    // Selects pivot and partition it.  
    int pivotInd = partition(fin, low, high);

    // Recursively partitions
    naive(fin, low, pivotInd - 1);
    naive(fin, pivotInd + 1, high);
  }

  /**
   * Sorts array by given low and high point.
   * Selects a pivot at the fin[low] and sorts fin based on the pivot.
   */
  public int partition(int[] fin, int low, int high) {
    int pivot = fin[low];
    int l = low;
    int h = high;

    while (l < h) {
      while (fin[l] <= pivot && l < h)
        l++;
      while (fin[h] > pivot)
        h--;
      if (l < h)
        swap(fin, l, h);
    }
    swap(fin, low, h);

    // returns pivot index
    return h;
  }

  /**
  * Function to swap two numbers in an array
  */
  public void swap(int[] fin, int a, int b) {
    int tmp = fin[a];
    fin[a] = fin[b];
    fin[b] = tmp;
  }
}

File: Python/quick_sort.py

''' QuickSort '''


def quickSort_naive(fin, low, high):
    if low < high:
        split_point = partition(fin, low, high)
        quickSort_naive(fin, low, split_point - 1)
        quickSort_naive(fin, split_point + 1, high)


def partition(fin, low, high):
    pivot = fin[low]

    left = low + 1
    right = high

    sort_done = False
    while not sort_done:
        while left <= right and fin[left] <= pivot:
            left += 1
        while fin[right] >= pivot and right >= left:
            right -= 1
        if right < left:
            sort_done = True
        else:
            tmp = fin[left]
            fin[left], fin[right] = fin[right], tmp
    tmp = fin[low]
    fin[low], fin[right] = fin[right], tmp

    return right


def quickSort_list(fin):
    if len(fin) <= 1:
        return fin

    pivots = [x for x in fin if x == fin[0]]
    lesser = quickSort_list([x for x in fin if x < fin[0]])
    greater = quickSort_list([x for x in fin if x > fin[0]])

    return lesser + pivots + greater


def quickSort_filter(fin):
    if len(fin) <= 1:
        return fin

    pivots = list(filter(lambda x: x == fin[0], fin))
    lesser = quickSort_list(list(filter(lambda x: x < fin[0], fin)))
    greater = quickSort_list(list(filter(lambda x: x > fin[0], fin)))

    return lesser + pivots + greater

